libnmstate/nm/connection.py

```diff
@@ -259,19 +259,26 @@ def waitfor_active_connection_async(self, ac):
     def _waitfor_active_connection_callback(
         self, _nm_act_con, _state, _reason, ac
     ):
-        ac.refresh_state()
+        cur_nm_act_conn = get_device_active_connection(self.nmdevice)
+        if cur_nm_act_conn and cur_nm_act_conn != ac.nm_active_connection:
+            logging.debug(
+                'Active connection of device {} has been replaced'.format(
+                    self.devname
+                )
+            )
+            ac.remove_handlers()
+            ac = ActiveConnection(cur_nm_act_conn)
+            self.waitfor_active_connection_async(ac)
```

Comment:
The function is called `waitfor_active_connection_async()` (taking an ac argument), but it doesn't really wait for the active connection... it waits for the device to reach an activated state.

Also, since it now swaps the active connection with another one, it does not ensure that the now activated profile is indeed the one that was intended.

I think the code should watch (subscribe to signals) primarily to the device's state.

Reply:
The name is indeed confusing, but in practice it waits for the ac to reach the activated state. If it it is not in "sync" with the device it pointed, it is re-triggered with an updated ac.
The ac of a device may change, but in practice, if I understand correctly, even the device may change (virtual ones). I mean in terms of their identifier/object.